tortoise/tortoise.go

```diff
@@ -636,268 +449,264 @@ func (t *turtle) getBallotBeacon(ballot *types.Ballot, logger log.Log) ([]byte,
-		// outcomes:
-		// 1. record our opinion on this block and go on evaluating the rest of the blocks in this layer to see if
-		//    we can verify the layer (if local and global consensus match, and global consensus is decided)
-		// 2. keep waiting to verify the layer (if not, and the layer is relatively recent)
-		// 3. trigger self healing (if not, and the layer is sufficiently old)
-		consensusMatches := globalOpinionOnBlock == localOpinionOnBlock
-		globalOpinionDecided := globalOpinionOnBlock != abstain
-
-		// If, for any block in this layer, the global opinion (summed block votes) disagrees with our vote (the
-		// input vector), or if the global opinion is abstain, then we do not verify this layer. This could be the
-		// result of a reorg (e.g., resolution of a network partition), or a malicious peer during sync, or
-		// disagreement about Hare success.
-		if !consensusMatches {
-			logger.With().Warning("global opinion on block differs from our vote, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
-		}
+		previous := t.verified
+		t.verified = t.verifying.verifyLayers(logger)
 
-		// There are only two scenarios that could result in a global opinion of abstain: if everyone is still
-		// waiting for Hare to finish for a layer (i.e., it has not yet succeeded or failed), or a balancing attack.
-		// The former is temporary and will go away after `zdist' layers. And it should be true of an entire layer,
-		// not just of a single block. The latter could cause the global opinion of a single block to permanently
-		// be abstain. As long as the contextual validity of any block in a layer is unresolved, we cannot verify
-		// the layer (since the effectiveness of each transaction in the layer depends upon the contents of the
-		// entire layer and transaction ordering). Therefore we have to enter self healing in this case.
-		// TODO: abstain only for entire layer at a time, not for individual blocks (optimization)
-		//   see https://github.com/spacemeshos/go-spacemesh/issues/2674
-		if !globalOpinionDecided {
-			logger.With().Warning("global opinion on block is abstain, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
+		if t.verified.After(previous) {
+			if err := persistContextualValidity(logger,
+				t.bdp,
+				previous, t.verified,
+				t.blocks,
+				t.localOpinion,
+			); err != nil {
+				return err
+			}
+		}
+		t.updateHistoricallyVerified()
+		if t.verified == t.processed.Sub(1) {
+			return nil
+		}
+	}
+	// keep votes starting from the layer when verifying tortoise failed to make progress
+	// for example, if verifying failed while counting ballots from layer 13 we will
+	// most likely have to use those ballots later in full mode.
```

Comment:
the comment here doesn't seem to match code in `keepVotes()`

say verified is 10 and processed is 12. we are only keeping votes for layer 12 but not 11. is that ok?

Reply:
i think it matches, maybe i need to improve wording though
layer 12 verifies layer 11, so if verifying failed we will use votes from layer 12 later in full mode.